### PR TITLE
Update neml2

### LINKED ISSUE #27483: Enable implicit conversion between `FileName` and `std::filesystem::path`
### Motivation

`FileName` subclasses `std::string`, and so it will be two layers of implicit conversion from `FileName` to `std::filesystem::path`. Functions accepting `std::filesystem::path` will not take `FileName`, which seems counter-intuitive to me.

### Design

Add the implicit conversion. But we need to be careful with the modification to the macro though, as we don't want other string-derived types like `VariableName` to be implicitly convertible to path.

### Impact

Happier coders. We can do
```cpp
foo(filename)
```
instead of
```cpp
foo(std::string(filename))
```

### LINKED ISSUE #27503: Update NEML2 submodule to tag v1.4.0
### Motivation

Bring in new features and bug fixes from NEML2.

### Design

Perform a NEML2 submodule update. Investigate any failures.

### Impact

New NEML2 submodule.

### PR DESCRIPTION
**Pull request by hugary1995**
Update NEML2 to v1.4.0 -- closes #27503

### CODE REVIEW

File: framework/include/utils/MooseTypes.h [hunk 1]
```diff
@@ -943,39 +959,45 @@ struct enable_bitmask_operators<Moose::RelationshipManagerType>
...
+ * operator to std::filesystem::path
+ */
+#define DerivativePathLikeStringClass(TheName)                                                     \
+  DerivativeStringClass(TheName##Base);                                                            \
+  class TheName : public TheName##Base                                                             \
   {                                                                                                \
-    static void to_json(json & j, const TheName & v);                                              \
+  public:                                                                                          \
+    using TheName##Base::TheName##Base;                                                            \
+    operator std::filesystem::path() const { return std::string(*this); }                          \
```
Comment (dschwen) [inline framework/include/utils/MooseTypes.h:974]: :+1:

Comment (dschwen) [reply framework/include/utils/MooseTypes.h:974]: Uhm, looks like this operator is not used for an implicit conversion though.

File: framework/include/utils/MooseTypes.h [hunk 2]
```diff
@@ -924,6 +925,21 @@ struct enable_bitmask_operators<Moose::RelationshipManagerType>
...
+ * This Maco is used to generate forward declarations for to_json methods for each of the
+ * string-derived class.
+ */
+#define DerivativeStringToJSONDecl(TheName)                                                        \
+  namespace nlohmann                                                                               \
+  {                                                                                                \
+  template <>                                                                                      \
+  struct adl_serializer<TheName>                                                                   \
+  {                                                                                                \
+    static void to_json(json & j, const TheName & v);                                              \
```
Comment (loganharbour) [inline framework/include/utils/MooseTypes.h:938]: https://github.com/hugary1995/moose/blob/8428f040c912cb1b59511b3a0ec6e7c51057d8af/framework/src/utils/MooseTypes.C#L33 ?

Comment (loganharbour) [reply framework/include/utils/MooseTypes.h:938]: oh, I see it now. this diff is confusing me

File: framework/include/utils/MooseTypes.h (hunk 1 above)
Comment (hugary1995) [reply framework/include/utils/MooseTypes.h]: fine...... I dropped this change.

File: modules/solid_mechanics/contrib/neml2_moose/neml2/base/config.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * TODO: make the following constants configurable
+ */
+/// The machine precision
+#define NEML2_EPS 1E-15
+/// The tolerance used in various algorithms
+#define NEML2_TOL 1E-6
+/// A tighter tolerance used in various algorithms
+#define NEML2_TOL2 1e-12
```
Comment (dschwen) [inline modules/solid_mechanics/contrib/neml2_moose/neml2/base/config.h:42]: those should be adjusted if `NEML2_DTYPE` get changed to `torch::kFloat32` for example, right?

**Review by dschwen — APPROVED** [review]

**Comment by dschwen** [comment]
Internal app tests are due to a temporary outage

File: modules/solid_mechanics/contrib/neml2_moose/neml2/base/config.h (hunk 3 above)
Comment (hugary1995) [reply modules/solid_mechanics/contrib/neml2_moose/neml2/base/config.h:42]: That's correct. We just haven't fully thought through how best to handle all the constants yet. They might all be run-time constants in the future.

**Pull request merged**